### PR TITLE
Topic recognition - s48008361 Siamese Submission

### PR DESCRIPTION
**Pull request by TheNoZER0**
This pull request incorporates a solution to classify skin lesions and identify malignant melanoma cases using a Siamese-based classifier network on the ISIC 2020 Challenge Dataset. The Siamese-based classifier allows for the recognition of malignant melanoma lesion images and benign lesion images. The primary objective was to classify these two cases with an accuracy greater than 80% through the use of triplet loss. AUC-ROC was used as the primary metric to evaluate the accuracy, achieving a score of 0.89, which exceeds the target.

**Highlights:**
-  Implementation of the `ISIC2020Dataset` class for generating triplets of positive, negative, and anchor images to be fed into the training loop
- Development of a Siamese Network using a pre-trained ResNet50 feature extractor with unfrozen layers for benign and malignant classification
- Application of data augmentation techniques to improve the model's robustness and generalization
- Comprehensive analysis of results using a confusion matrix, AUC-ROC curve, and visualization of t-SNE embeddings to evaluate the model's performance on the test set
- Addressing the challenge of class imbalance in the ISIC dataset through oversampling and balanced batch sampling

**Contents / Files Added:**
- `dataset.py`: Includes `ISIC2020Dataset` and `BalancedBatchSampler` classes and `preprocess_dataset` helper for preprocessing data into benign and malignant folders, 80/20 train-test split, and data augmentations to ensure balanced batches and training set
- `modules.py`: Defines the architecture for a ResNet50 pre-trained feature extractor, Siamese Network, Triplet Loss, and the Classifier
- `train.py`: Implements the Siamese Network training loop and logic, loss graphs, and accuracy graphs visualization as well as classifier training
- `predict.py`: Generates visualization of confusion matrix, AUC-ROC curve, t-SNE embeddings, and prints results and statistics of the model

**Requirements for setup:**
- PyTorch 2.1.1+cu11.8 (CUDA 11.8)
- Python 3.8.19
- Other dependencies include matplotlib, numpy, torchvision, scikit-learn, seaborn, pandas, tqdm, pillow
- Detailed setup and installation instructions are included in the README

Thanks,
Zain Al-Saffi (s4800836)

### CODE REVIEW

**Comment by hanemma7moud** [comment]
### This is an initial inspection, no action is required at this point
Recognition Problem : total : **20**

Solves problem: The solution is well-suited to the problem (**4.5**).
 (-0.5 for not achievingthe 80% accuracy)
A good performance in detecting malignant skin lesions, with a strong AUC-ROC and high sensitivity towards malignant cases. Accuracy =76.9%, Sensitivity = 88.8%, and Specificity= 76.7%
Triplet loss is utilized, which is suitable for the problem.
Class imbalance handling is achieved through oversampling of the minority class, balanced batch sampling, and a stratified train-validation split.
Confusion matrix, AUC-ROC curve, and t-SNE visualization provide insights into performance, sensitivity, and separability of learned features.

Implementation functions as intended (**3**)

Good design: Well-designed (**1**)

Commenting: Clear and sufficient comments throughout the code. (**1**)

Difficulty: Hard (**10**)

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Good design and implementation. </td><td></td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>Header blocks. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>Good solution to problem. </td><td></td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td></td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting present. </td><td></td></tr>
	<tr><td></td><td>No Data leakage found. Has data splits. </td><td></td></tr>
	<tr><td></td><td>Difficulty : Hard.  Siamese network (Hard Difficulty)</td><td></td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Good Meaningful commit messages. </td><td></td></tr>
	<tr><td></td><td>Good Progressive commits. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Good. </td><td></td></tr>
	<tr><td></td><td>Model/technical explanation :Good. </td><td></td></tr>
	<tr><td></td><td>Description and Comments :Good. </td><td></td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. PDF Checked.</td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch). On time.</td><td></td></tr>
	<tr><td></td><td>No Feedback required. </td><td></td></tr>
	<tr><td></td><td>Request Description is good. </td><td></td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>0</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_

**Comment by shakes76** [comment]
-1 for not achieving main accuracy. Great work otherwise.
